keps/sig-apps/2214-indexed-job/README.md

```diff
@@ -0,0 +1,487 @@
+```
+
+### Notes/Constraints/Caveats (Optional)
+
+* An earlier proposal for [indexed Job]
+suggested the support for work lists, i.e. passing different parameters to each
+Pod. We decided to leave this out of the proposal to keep it simple and
+because work lists can be implemented in a startup script using the completion
+index as building block.
+* The semantics of an indexed Job are similar to a StatefulSet, in the sense
+that Pods have an associated index and Pods.
+However, the APIs have a major difference: a StatefulSet doesn't have completion
+semantics, as opposed to Jobs.
+
+[indexed Job]: https://github.com/kubernetes/community/blob/b21d1b27c8c748bf81283c2d89cde2becb5f2709/contributors/design-proposals/apps/indexed-job.md
+
+### Risks and Mitigations
+
+Jobs have a known issue in which more than one Pod can be started even if
+parallelism and completion are set to 1 ([reference]). In the case of indexed
+Jobs, this translates as more than one Pod having the same index.
+
+Just like for existing Job patterns, workloads have to handle duplicates at the
+application level.
+
+[reference]: https://kubernetes.io/docs/concepts/workloads/controllers/job/#handling-pod-and-container-failures
+
+## Design Details
+
+### Indexed Jobs
+
+Indexed Jobs are Jobs [with fixed completion count], that is, Jobs that have a
+non-nil `.spec.completions` field.
+
+### Pod detail
+
+The Pod and PodSpec APIs don't get any new fields. However, Pods created for
+Indexed Jobs get the annotation `alpha.batch.k8s.io/job-completion-index` with a
+value equal to its completion index. The annotation is immutable.
+
+The annotation can be accessed through the downward API as a file or environment
+variable.
+
+### Job completion and restart policy
+
+When dealing with Indexed Jobs, the Job controller keeps track of Pod
+completions for each index from 0 to `.spec.completions - 1`. When a Pod
+finishes successfully, the controller doesn't create another Pod with the same
+completion index. The Job controller doesn't remove succeeded or failed Pods.
+
+If an entire Pod fails, such as when the Pod gets kicked off the node or if a
+container of the Pod fails and `.spec.template.spec.restartPolicy = "Never"`,
+the Job controller gets the completion index for the failed Pod and creates a
+new Pod with the same index. The application needs to handle restarts in a
+different Pod.
+
+The kubelet handles container restarts as usual, according to the
+`spec.template.spec.restartPolicy`.
+
+### Job scaling
+
+Jobs can be scaled up and down by varying `.spec.parallelism` (note that
+`.spec.completions` is an immutable field).
+
+When starting or scaling up a Job, the Job controller creates Pods with lower
+completion index first, as long as there is no other completed or running Pod
+with the same index. This is to make the controller behavior more
+predictable. We do not offer any guarantees.
+
+Scale down is unaffected by completion index.
+
```

Comment:
Can you show how `.status` changes and how `kubectl describe` changes?

In the case of an indexed job, I think something like this would be nice:
```
$ kubectl describe jobs/parallel-work
Name:           countdown
Namespace:      default
...
Parallelism:    100
Completions:    10
Pods Statuses:  10 Running / 27 Succeeded / 0 Failed
Completed Indexes: [1-25,27-29]
```

While this list could get somewhat long, it could be range-compressed like this: https://stackoverflow.com/questions/2270910/how-to-reduce-consecutive-integers-in-an-array-to-hyphenated-range-expressions

Reply:
Other than the output format, we still need to add the full list of completed indexes to the Status. Is that acceptable?

Reply:
Maybe I'm too paranoid, but I can easily imagine case of someone who wants to run 1M indexes.
And in some pathological scenarios where e.g. the app itself is crashing for even numbers, we can end up with
Completed Indexes: [1, 3, 5, 7, ...., 999999]
Which we can't really compress.

I'm fine with hearing "you're stupid - this will not happen in practise", but it's something lights the yellow lamp for me.

That said - I fully agree that it would be useful and doable in majority of cases.
So maybe we can do something in between, which is:
- try to compress it
- if it's too long, just cut it to e.g. 1k bytes (or whatever other number), so something like (for above pathological case):
- 
`Completed Indexes: [1, 3, 5, <the next e.g. 400 odd numbers or so>, 799, <truncated> ]`

Reply:
Hmm - that actually made me think if we don't really want to store it, because on controller restart and many of finished pod already GC-ed we would actually lose the knowledge about them...

On a second thought to the above - the number of segments can probably be bounded by `spec.parallelism`, so how large in practice that can be...

I guess I've just lost myself in all those thoughts - I think I need to think more about this.

Reply:
In the topic of GC kubernetes/kubernetes#28486

Reply:
I might do something like this:  

- If there are fewer than 1000 completions, then set the status field.  
    - Worst case is 500 (only odd ones) indexes complete  times 4 bytes per (e.g. `997,`) = 2kB which is no big deal, considering 1.5MB object size limit
- If completions > 1000, never set this particular status field.
  - Status is always initially empty, so clients should expect to sometimes see fields missing.
  - Clients should wait on the condition, not on random status fields.
- Document that setting this status field is not supported for large completion counts.
-  Limit frequency of status updates to be nice to api server.  = 500, 3 dig object size limit is 1.5MB. 
- Consider raising in a few releases if there is demand.

Reply:
Added as unresolved, also targetting Beta.

Reply:
I like the idea of using `status` as storage to solve the "forgot which pods we already ran" problem.   I think @wojtek-t  is alluding to that.

However, I think you can solve that in a separate subsequent KEP.

Reply:
> I like the idea of using status as storage to solve the "forgot which pods we already ran" problem. I think @wojtek-t is alluding to that.

Yes - that was something i wanted us to think about that.
But I agree with you it probably shouldn't be this KEP - it should probably be a separate feature.

Reply:
#28486 your input would be appreciated there. I'm about to initiate a KEP on it.

Reply:
Okay, I looked at that issue.

Seems like you can use the same method for indexed Job, except that you need to update the `status.completedIndexes` atomically in step 5 too.  But that means `status.completedIndexes` always needs to get set.

We can put a tighter bound on its size if we limit parallelism.  We don't need to limit completions.
`status.completedIndexes` has format `a-b,c-d,e-f,...`.  The number of commas cannot be more than the max parallelism allowed by the system for a job. (say 1000).
The length of one list item with comma e.g. `a-b,` is `2 * log10(max_completions) + 2`.
So the max string length of `status.completedIndexes` is `max_parallelism *  (2 * log10(max_completions) + 2)`

If you set `max_parallelism = 10^3`
and you set `max_completions = 10^9`
then max length of `status.completedIndexes` is  20kB.

Reply:
"640K is more memory than anyone will ever need on a computer" -- (probably not) Bill Gates

Should we really limit parallelism way bellow what a k8s cluster can support?

Reply:
What are you thinking as an alternative?

Reply:
I can think of 2:
- To not keep track of indexes in status (and keep around the successful Pods with a finalizer until the entire Job succeeds)
- Have a limit that is more future proof. We know k8s can run up to 15k nodes. Maybe 10^5 or 10^6 would be reasonable.

Reply:
Blocking deletion of lots of pods, when there are lots of users running concurrent jobs, might overload the apiserver? 
Maybe @wojtek-t can be more specific on what limits should be here.

A limit of 15k pods parallelism and 100k indices would take up would take 15k * (2*5 + 2) = 180k bytes.
With 80 char lines, getting the pod status will take over 2000 lines, which is not a great experience.

Reply:
If using `kubectl describe` we can probably stop showing the indexes at a lower threshold.

Reply:
One option could be to separate the presentation and the state storage into separate fields..
* For presentation purposes, show `status.completedIndexes` when it is less than than `x` characters long, else do not set. Do this in this KEP
* For state storage,  to address kubernetes/kubernetes#28486 without delaying pod deletion, one can store the state of completed pods in a base64 encoded bitmask, e.g. `status.completionBitmask`.  10^6 bits fits into 16k chars, which is 200 lines at 80 chars per line.  Annoying but better. 
Once this KEP and kubernetes/kubernetes#28486 are implemented, both fields would be present in indexedjobs.

Reply:
> For presentation purposes, show status.completedIndexes when it is less than than x characters long, else do not set. Do this in this KEP

What about calculating the string to display from the bitmask on kubectl?

Reply:
See latest commit